proposals/1412/client_proxy.md

```diff
@@ -0,0 +1,157 @@
+    SEND lineage_event TO BACKEND
+    checkpoint(partition, offset)
+    offset += 1
+
+/**
+ * Blocking read which wait until the event is available.
+ * It might be implemented with retry and sleep.
+ */
+FUNC read(partition, offset):
+    return `SELECT event, continue FROM partition WHERE rowid=offset`
+```
+
+##### Partition GC: runs periocally
+```
+/* Delete partitions that are already processed */
+checkpointed_partition_id = `SELECT partition_id FROM checkpoint`
+deletion_candidates = `SELECT * FROM partitions WHERE rowid < checkpointed_partition_id`
+delete_partitions(deletion_candidates)
+
+/* Delete partitions that are older than limit hour */
+deletion_candidates = `SELECT * FROM partitions WHERE created_at < NOW() - LIMIT HOURS`
+delete_partitions(deletion_candidates)
+
+/* Delete oldest partitions until the total size is smaller then limit size */
+deletion_candidates = list()
+current_total_size = `SELECT SUM(size) FROM partitions`
+partition_records = `SELECT * FROM partitions ORDER BY rowid ASC`
+FOR partition_record IN partition_records:
+    if current_total_size < LIMIT SIZE:
+        BREAK
+    deletion_candidates.append(partition_record)
+    current_total_size -= partition_record.size
+delete_partitions(deletion_candidates)
+
+FUNC delete_partitions(deletion_candidates):
+    FOR partition IN deletion_candidates:
+        DROP TABLE partition IF EXISTS
+        DELETE FROM partitions WHERE rowid=partition.rowid
+```
+
+# Future Milestone
+We might add Kafka transport for the consumer so that it can send to not only the lineage API backend but also the kafka topic. With this implemented, we are able to use this proxy as a replacement of the backend proxy. The major job of both of proxy is receiving and forwarding events, which is quite an overlap, and we could be able to replace the proxy backend with this, like just one proxy implementation that can be used as either a client proxy or server side reverse proxy. One missing piece of this design for that is ability to forward events to the multiple destination that the proxy backend can do. Having multiple destination will require the change of the design in order to deal with the outage of subset of backends which will complicate the design, and I will put this as a future work.
+
+# Appendix
+- [Gin](https://github.com/gin-gonic/gin) is a web framework written in Go
+- [Lumberjack](https://github.com/natefinch/lumberjack) is a logging library that support rotation based on size and ages
+- [Tail](https://github.com/hpcloud/tail) is a go library for log tailing that is designed to work with a log rotation
+- [Sqlite](https://www.sqlite.org/) is a database engine that strives to provide local data storage for applications
+
+# Side Note
+As this proposes a client side proxy, which will make a confusion with existing Kafka proxy, I would like to propose rename the Kafka proxy to the reversed proxy, as it sits in front of the backend.
```

Comment:
Do you mean the `proxy` [backend](https://github.com/OpenLineage/OpenLineage/tree/main/proxy/backend)?

Reply:
Yes. Let me make it clear.